### PR TITLE
[location] Add support for user-initiated background tracking without background permission

### PR DESCRIPTION
**Pull request by byCedric**
# Why

Google has restricted the `ACCESS_BACKGROUND_LOCATION` permission and added an extra review when using this permission. We already made it an opt-in, but Google has split the location usage into [two main use cases](https://developer.android.com/training/location/permissions#types-of-access), as described below.

Right now, we only support the "background" and not the "foreground" use case. We already have the `foregroundService` option, so we should support it without any user-facing API changes. See [this issue thread for community-provided scenarios with the foreground use case](https://github.com/expo/expo/issues/11854).

1. **The ["foreground" use case](https://developer.android.com/training/location/permissions#foreground)**
> If your app contains a feature that shares or receives location information only once or for a defined amount of time, then that feature requires foreground location access. Some examples include the following:
>
> - Within a navigation app, a feature allows users to get turn-by-turn directions.
> - Within a messaging app, a feature allows users to share their current location with another user.
>
>The system considers your app to be using foreground location if a feature of your app accesses the device's current location in one of the following situations:
>
> - An activity that belongs to your app is visible.
> - Your app is running a foreground service. When a foreground service is running, the system raises user awareness by showing a persistent notification. Your app retains access when placed in the background, such as when the user presses the Home button on their device or turns their device's display off.

2. **The ["background" use case](https://developer.android.com/training/location/permissions#background)**
> An app requires background location access if a feature within the app constantly shares location with other users or uses the Geofencing API. Several examples include the following:
> 
> - Within a family location sharing app, a feature allows users to continuously share location with family members.
> - Within an IoT app, a feature allows users to configure their home devices such that they turn off when the user leaves their home and turn back on when the user returns home.
>
> The system considers your app to be using background location if it accesses the device's current location in any situation other than the ones described in the foreground location section.

# How

- **Added a public static `useForegroundService` method in `LocationTaskConsumer`**
This checks if `foregroundService` is defined in the options. By adding it here, we can reuse the constant from this class in the module.
- **Added a check in the `startLocationUpdatesAsync` to determine if we should validate the background permission**
This helps us not-blocking the "foreground" use case. It only blocks when `foregroundService` option and the background permission are NOT defined.
- **Removed additional background permission checks in the `stopLocationUpdatesAsync` and `hasStartedLocationUpdatesAsync`**
Because the validation is now dependent on the provided options, we can't validate this here. I think it still makes sense to allow developers to check/stop, even if they misconfigured it. It still won't start the background location without an error when misconfigured.

# Test Plan

- Clone my [example app "Office Marathon"](https://github.com/bycedric/office-marathon)
- Run `$ expo eject`
- Remove the `ACCESS_BACKGROUND_LOCATION` permission from manifest + `app.json`
- Add this PR manually to the local `expo-location`.

I tested this app with this change for the following scenarios:

Scenario | Android 30 | Android 29 | Android 28
--- | --- | --- | ---
App foregrounded | ✅ | ✅ | ✅
App foregrounded with locked screen | ✅ | ✅ | ✅
App backgrounded with locked screen | ✅ | ✅ | ✅

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-location/android/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -8,6 +8,7 @@
   <application>
     <service
       android:name=".services.LocationTaskService"
-      android:exported="false" />
+      android:exported="false"
+      android:foregroundServiceType="location" />
```
Comment (byCedric) [inline packages/expo-location/android/src/main/AndroidManifest.xml:12]: Android automatically uses the "manifest service type" when starting the foreground service. It's possible to add more, and start the service for specific types. See [this documentation](https://developer.android.com/guide/components/foreground-services#loc-camera)

File: packages/expo-location/android/src/main/java/expo/modules/location/taskConsumers/LocationTaskConsumer.java [hunk 2]
```diff
@@ -159,6 +159,10 @@ public void onFinished(Map<String, Object> response) {
     return true;
   }
 
+  public static boolean useForegroundService(Map<String, Object> options) {
```
Comment (byCedric) [inline packages/expo-location/android/src/main/java/expo/modules/location/taskConsumers/LocationTaskConsumer.java:162]: I'm not sure if this is common. I did this to keep the `FOREGROUND_SERVICE_KEY` "in sync" with validation from the module file. We can also:
1. Copy the string/constant to the module
2. Expose this string/constant to the module

**Comment by byCedric** [comment]
Unfortunately, for geofencing, we still need the background permission. There is no counterpart for this API I think.

File: packages/expo-location/android/src/main/java/expo/modules/location/taskConsumers/LocationTaskConsumer.java (hunk 2 above)
Comment (lukmccall) [reply packages/expo-location/android/src/main/java/expo/modules/location/taskConsumers/LocationTaskConsumer.java:162]: Actually, I like this ;)

**Review by lukmccall — APPROVED** [review]
LGTM 💪💪💪

I think it's worth adding something to our documentation that explains google restrictions. However, if we currently have something like this, just ignore my comment 😅

**Comment by byCedric** [comment]
No, you are right! Our docs currently don't reflect or communicate this. We have to redo the permissions doc anyways, lets add this to linear.

**Comment by arabold** [comment]
I patched my local checkout of `expo-location` 12.0.2 with this and it immediately worked. Now I'm receiving location updates even while the app is "minimized" without the need for background permission. Great work! I'm very excited as this simplifies my whole setup dramatically compared to @mauron85/react-native-background-geolocation which I used before.

**Review by tsapeta — APPROVED** [review]
Looks good, thanks for taking care of these modules! 🙇‍♂️

File: packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java [hunk 3]
```diff
@@ -463,7 +463,12 @@ public void hasServicesEnabledAsync(final Promise promise) {
 
   @ExpoMethod
   public void startLocationUpdatesAsync(String taskName, Map<String, Object> options, final Promise promise) {
-    if (isMissingBackgroundPermissions()) {
+    boolean useForegroundService = LocationTaskConsumer.useForegroundService(options);
+
+    // There are two ways of starting this service.
+    // 1. As a background location service, this requires the background location permission.
+    // 2. As a user-initiated foreground service with notification, this does NOT require the background location permission.
+    if (!useForegroundService && isMissingBackgroundPermissions()) {
```
Comment (byCedric) [inline packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java:471]: Renaming to `shouldUseForegroundService`:

```suggestion
    if (!shouldUseForegroundService && isMissingBackgroundPermissions()) {
```

File: packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java [hunk 4]
```diff
@@ -463,7 +463,12 @@ public void hasServicesEnabledAsync(final Promise promise) {
 
   @ExpoMethod
   public void startLocationUpdatesAsync(String taskName, Map<String, Object> options, final Promise promise) {
-    if (isMissingBackgroundPermissions()) {
+    boolean useForegroundService = LocationTaskConsumer.useForegroundService(options);
```
Comment (byCedric) [inline packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java:466]: Renaming to `shouldUseForegroundService`:

```suggestion
    boolean shouldUseForegroundService = LocationTaskConsumer.shouldUseForegroundService(options);
```

**Comment by rvmelo** [comment]
On expo-location version ~12.1.3 it still doesn't work.  I'm using expo sdk 42.0.0 and I get the following error when I try to track user location on background: **Attempt to invoke interface method 'java.util.Map o.e.c.a.d.getOptions()' on a null object reference** . Sometimes it works and other times it doesn't work.

**Pull request merged**